perf(ivy): chain styling instructions

crisbeto opened the pull request:
Adds support for chaining of `styleProp`, `classProp` and `stylePropInterpolateX` instructions whenever possible which should help generate less code. Note that one complication here is for `stylePropInterpolateX` instructions where we have to break into multiple chains if there are other styling instructions inbetween the interpolations which helps maintain the execution order.

matsko commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -664,8 +668,16 @@ function createHostBindingsFunction(
+        const calls: o.Expression[][] = [];
+
+        instruction.calls.forEach(call => {
+          totalHostVarsCount += Math.max(call.allocateBindingSlots - 1, 0);
```
Add a comment explaining why there's a `-1`

matsko commented on packages/compiler/src/render3/view/styling_builder.ts:
```diff
@@ -25,10 +24,14 @@ const IMPORTANT_FLAG = '!important';
-  sourceSpan: ParseSourceSpan|null;
   reference: o.ExternalReference;
-  allocateBindingSlots: number;
+  calls: StylingInstructionCall[];
```
Add a comment explaining what this new member is.

matsko commented on packages/compiler/src/render3/view/styling_builder.ts:
```diff
@@ -25,10 +24,14 @@ const IMPORTANT_FLAG = '!important';
  * A styling expression summary that is to be processed by the compiler
  */
 export interface StylingInstruction {
```
Since we're only using one property on this interface, why not just return a array of `StylingInstructionCall` instances?

crisbeto replied:
It's set up like this so that it's easier to group calls to the same instruction together. If we wanted to change it, we'd have to do something like `StylingInstructionCall[][]` instead, but that would mean looping through all the calls when we're trying to figure out whether to chain them.

matsko replied:
ah I see. I missed the `reference` member. This is fine.

matsko commented on packages/core/src/render3/instructions/styling.ts:
```diff
@@ -133,7 +135,7 @@ export function stylePropInternal(
  * @codeGenApi
  */
-export function ɵɵclassProp(className: string, value: boolean | null): void {
+export function ɵɵclassProp(className: string, value: boolean | null): TsickleIssue1009 {
```
Add a comment explaining the return value and why it's an instance of `TsickleIssue1009`

crisbeto replied:
Isn't this the point of the `TsickleIssue1009` type in the first place? Also Andrew mentioned that this isn't an issue anymore so as soon as these changes land I'll clean out all the usages of the `TsickleIssue1009` type.

matsko commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -1127,7 +1131,11 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
+          fnParams.push(...property.params);
+        }
         if (property.name) {
           fnParams.unshift(o.literal(property.name));
```
Add a comment explaining this `unshift` operation.

crisbeto commented:
The feedback has been addressed @matsko.

kara commented on packages/compiler-cli/test/compliance/r3_view_compiler_styling_spec.ts:
```diff
@@ -1461,6 +1456,351 @@ describe('compiler compliance: styling', () => {
+      expectEmit(result.source, template, 'Incorrect template');
+    });
+
+    it('should break into multiple chains if there are other styling instructions inbetween',
```
```suggestion
    it('should break into multiple chains if there are other styling instructions in between',
```

kara commented on packages/compiler-cli/test/compliance/r3_view_compiler_styling_spec.ts:
```diff
@@ -1461,6 +1456,351 @@ describe('compiler compliance: styling', () => {
+         expectEmit(result.source, template, 'Incorrect template');
+       });
+
+    it('should break into multiple chains if there are other styling interpolation instructions inbetween',
```
```suggestion
    it('should break into multiple chains if there are other styling interpolation instructions in between',
```

kara reviewed (approved):
LGTM aside from typos. Note that we should be careful with the G3 presubmit for this because we are changing binding order in some cases, which could break some G3 targets.

crisbeto committed:
perf(ivy): chain styling instructions

Adds support for chaining of `styleProp`, `classProp` and `stylePropInterpolateX` instructions whenever possible which should help generate less code. Note that one complication here is for `stylePropInterpolateX` instructions where we have to break into multiple chains if there are other styling instructions inbetween the interpolations which helps maintain the execution order.

matsko commented:
presubmit = http://test/OCL:281116837:BASE:281116883:1574105974370:89ff78f9

ivy presubmit = https://test.corp.google.com/ui#id=OCL:281116837:BASE:281117274:1574106065978:3e19703d
